レビューしました。

#108 の目的である「コピー操作にも #69 と同等の競合確認・進捗・キャンセルを入れる」方向性は良いです。`CopyItemsAsync` を `MoveItemsAsync` と同じ形に寄せ、内部 Ctrl+ドラッグと Ctrl+V コピー貼り付けの両方から `ShowCopyConflictDialogAsync` を渡す構成も自然です。

ただし、現時点では blocking が 2 点あります。

---

## 1. P1: 同一パスへのコピー上書きで元ファイル/フォルダーを削除し得ます

Codex の指摘どおり、`CopyItemsAsync` は `targetPath` が `sourcePath` と同一でも `ItemExistsAtPath(targetPath)` の競合処理に入り、ユーザーが「上書き」を選ぶと既存項目を削除してからコピーしようとします。

同一フォルダーへの Ctrl+ドラッグコピー、または同一フォルダーで Ctrl+C → Ctrl+V した場合に、コピー元とコピー先が同一パスになります。この状態で上書きを選ぶと、元ファイル/元フォルダーを削除してから `File.Copy(sourcePath, targetPath, overwrite: false)` / `CopyDirectory(...)` に進むため、データ喪失につながります。

これは必ず修正してください。

対応案:

```csharp
if (IsSamePath(sourcePath, targetPath))
{
    // 同一パスへのコピーは上書き不可。Explorer 風にするなら別名コピーは将来対応として、
    // 今回は安全側に倒して skip 扱いが妥当。
    skipCount++;
    progress?.Report(i + 1);
    continue;
}
```

少なくとも `ItemExistsAtPath(targetPath)` の上書き分岐に入る前に `IsSamePath(sourcePath, targetPath)` を検出し、削除処理へ進まないようにしてください。

追加テストも必要です。

- `CopyItemsAsync_SamePath_SkipsWithoutDeletingSource`
- フォルダー版も可能なら `CopyItemsAsync_SameFolderPath_SkipsWithoutDeletingSourceFolder`

---

## 2. CI failure: `StubFileOperationService` が `CopyItemsAsync` を実装していません

CI は Debug build で失敗しています。

```text
FileBrowserPaneViewModelTests.cs(1261,53): error CS0535:
'FileBrowserPaneViewModelTests.StubFileOperationService' does not implement interface member
'IFileOperationService.CopyItemsAsync(...)'
```

`IFileOperationService` に `CopyItemsAsync` を追加したので、テスト用 stub にも実装が必要です。既存の `MoveItemsAsync` と同様に、まずは `Task.FromResult(CopyItemsResult)` を返す形でよいと思います。

また、今回の ViewModel 変更を考えると、以下のようなテストがあると安心です。

- `ExecuteCopyItemsToFolderAsync_WithConflictCallback_UsesAsyncPath`
- `ExecutePasteAsync_CopyOperation_WithConflictCallback_UsesCopyAsyncPath`
- `ExecuteCopyItemsToFolderAsync_CancelledOnly_DoesNotShowErrorDialog` は View 側なので難しければ不要

---

## 補足

コピー進捗・コピー中止ボタン・完了サマリー・コピー競合ダイアログの追加自体は #108 の受け入れ条件に合っています。`ShowMoveConflictDialogAsync` と `ShowCopyConflictDialogAsync` はほぼ同型なので、将来的には共通化余地がありますが、今回の PR ではブロッカーではありません。

現時点では CI / E2E / Security Checks が failure ですが、CI は上記 build error で止まっているため、まず `StubFileOperationService` と同一パス上書きガードを直してから再確認したいです。

--- a/PhotoGeoExplorer.Tests/FileBrowserPaneViewModelTests.cs
+++ b/PhotoGeoExplorer.Tests/FileBrowserPaneViewModelTests.cs
@@ -1099,6 +1099,53 @@ public class FileBrowserPaneViewModelTests
     }
 
     [Fact]
+    public async Task ExecuteCopyItemsToFolderAsync_WithConflictCallback_ReturnsAsyncSummary()
+    {
+        var tempDir = CreateTempTestDirectory();
+        try
+        {
+            using var vm = CreateViewModelWithFakes(out var fakePaneService, out var stubOp);
+            stubOp.CopyItemsResult = new FileOperationSummary(2, 0, Array.Empty<FileOperationFailure>());
+            await vm.LoadFolderAsync(tempDir);
+            var before = fakePaneService.LoadFolderCallCount;
+            var items = new List<PhotoListItem> { CreatePhotoListItem("a.jpg"), CreatePhotoListItem("b.jpg") };
+
+            var summary = await vm.ExecuteCopyItemsToFolderAsync(
+                items, tempDir, (_, _) => Task.FromResult(ConflictResolution.Skip));
+
+            Assert.Equal(2, summary.SuccessCount);
+            Assert.Equal(before + 1, fakePaneService.LoadFolderCallCount);
+        }
+        finally
+        {
+            CleanupTempDirectory(tempDir);
+        }
+    }
+
+    [Fact]
+    public async Task ExecutePasteAsync_CopyOperation_WithConflictCallback_ReturnsSuccess()
+    {
+        var tempDir = CreateTempTestDirectory();
+        try
+        {
+            using var vm = CreateViewModelWithFakes(out _, out var stubOp);
+            stubOp.CopyItemsResult = new FileOperationSummary(1, 0, Array.Empty<FileOperationFailure>());
+            await vm.LoadFolderAsync(tempDir);
+            vm.SetClipboard(new List<PhotoListItem> { CreatePhotoListItem("photo.jpg") }, ClipboardOperation.Copy);
+
+            var summary = await vm.ExecutePasteAsync(
+                resolveMoveConflictAsync: null,
+                resolveCopyConflictAsync: (_, _) => Task.FromResult(ConflictResolution.Skip));
+
+            Assert.Equal(1, summary.SuccessCount);
+        }
+        finally
+        {
+            CleanupTempDirectory(tempDir);
+        }
+    }
+
+    [Fact]
     public async Task ExecutePasteAsync_CutOperation_CallsMoveItemsAndClearsClipboard()
     {
         var tempDir = CreateTempTestDirectory();
@@ -1315,6 +1362,12 @@ public class FileBrowserPaneViewModelTests
             IProgress<int>? progress = null,
             CancellationToken cancellationToken = default) => Task.FromResult(MoveItemsResult);
         public FileOperationSummary CopyItems(IReadOnlyList<PhotoListItem> items, string destinationFolder) => CopyItemsResult;
+        public Task<FileOperationSummary> CopyItemsAsync(
+            IReadOnlyList<PhotoListItem> items,
+            string destinationFolder,
+            Func<string, bool, Task<ConflictResolution>> resolveConflictAsync,
+            IProgress<int>? progress = null,
+            CancellationToken cancellationToken = default) => Task.FromResult(CopyItemsResult);
         public Task<FileOperationSummary> DeleteItemsAsync(IReadOnlyList<PhotoListItem> items) => Task.FromResult(DeleteItemsResult);
     }
 

--- a/PhotoGeoExplorer.Tests/FileOperationServiceTests.cs
+++ b/PhotoGeoExplorer.Tests/FileOperationServiceTests.cs
@@ -801,6 +801,119 @@ public sealed class FileOperationServiceTests
     }
 
     // =========================================================
+    // CopyItemsAsync
+    // =========================================================
+
+    [Fact]
+    public async Task CopyItemsAsync_SamePath_SkipsWithoutDeletingSource()
+    {
+        var tempDir = CreateTempTestDirectory();
+        try
+        {
+            var srcPath = Path.Combine(tempDir, "photo.jpg");
+            await File.WriteAllTextAsync(srcPath, "content");
+            var items = new List<PhotoListItem> { CreateFileItem(srcPath) };
+
+            // 同一フォルダへのコピーは同一パスになる
+            var summary = await _service.CopyItemsAsync(
+                items, tempDir, (_, _) => Task.FromResult(ConflictResolution.Overwrite));
+
+            Assert.Equal(0, summary.SuccessCount);
+            Assert.Equal(1, summary.SkipCount);
+            Assert.True(File.Exists(srcPath), "ソースファイルが削除されていないこと");
+        }
+        finally
+        {
+            CleanupTempDirectory(tempDir);
+        }
+    }
+
+    [Fact]
+    public async Task CopyItemsAsync_ConflictOverwrite_OverwritesExisting()
+    {
+        var tempDir = CreateTempTestDirectory();
+        var destDir = CreateTempTestDirectory();
+        try
+        {
+            var srcPath = Path.Combine(tempDir, "photo.jpg");
+            var conflictPath = Path.Combine(destDir, "photo.jpg");
+            await File.WriteAllTextAsync(srcPath, "new");
+            await File.WriteAllTextAsync(conflictPath, "old");
+            var items = new List<PhotoListItem> { CreateFileItem(srcPath) };
+
+            var summary = await _service.CopyItemsAsync(
+                items, destDir, (_, _) => Task.FromResult(ConflictResolution.Overwrite));
+
+            Assert.Equal(1, summary.SuccessCount);
+            Assert.Equal("new", await File.ReadAllTextAsync(conflictPath));
+            Assert.True(File.Exists(srcPath), "コピー元は残っていること");
+        }
+        finally
+        {
+            CleanupTempDirectory(tempDir);
+            CleanupTempDirectory(destDir);
+        }
+    }
+
+    [Fact]
+    public async Task CopyItemsAsync_ConflictSkip_LeavesExistingAndCounts()
+    {
+        var tempDir = CreateTempTestDirectory();
+        var destDir = CreateTempTestDirectory();
+        try
+        {
+            var srcPath = Path.Combine(tempDir, "photo.jpg");
+            var conflictPath = Path.Combine(destDir, "photo.jpg");
+            await File.WriteAllTextAsync(srcPath, "new");
+            await File.WriteAllTextAsync(conflictPath, "old");
+            var items = new List<PhotoListItem> { CreateFileItem(srcPath) };
+
+            var summary = await _service.CopyItemsAsync(
+                items, destDir, (_, _) => Task.FromResult(ConflictResolution.Skip));
+
+            Assert.Equal(0, summary.SuccessCount);
+            Assert.Equal(1, summary.SkipCount);
+            Assert.Equal("old", await File.ReadAllTextAsync(conflictPath));
+        }
+        finally
+        {
+            CleanupTempDirectory(tempDir);
+            CleanupTempDirectory(destDir);
+        }
+    }
+
+    [Fact]
+    public async Task CopyItemsAsync_Cancelled_StopsProcessing()
+    {
+        var tempDir = CreateTempTestDirectory();
+        var destDir = CreateTempTestDirectory();
+        try
+        {
+            var src1 = Path.Combine(tempDir, "a.jpg");
+            var src2 = Path.Combine(tempDir, "b.jpg");
+            await File.WriteAllTextAsync(src1, "a");
+            await File.WriteAllTextAsync(src2, "b");
+            var items = new List<PhotoListItem> { CreateFileItem(src1), CreateFileItem(src2) };
+
+            using var cts = new CancellationTokenSource();
+            await cts.CancelAsync();
+
+            var summary = await _service.CopyItemsAsync(
+                items, destDir, (_, _) => Task.FromResult(ConflictResolution.Skip),
+                cancellationToken: cts.Token);
+
+            Assert.Equal(0, summary.SuccessCount);
+            Assert.True(summary.HasFailures);
+            Assert.Equal(FileOperationError.Cancelled, summary.Failures[0].Error);
+        }
+        finally
+        {
+            CleanupTempDirectory(tempDir);
+            CleanupTempDirectory(destDir);
+        }
+    }
+
+    // =========================================================
     // DeleteItemsAsync
     // =========================================================
 

--- a/PhotoGeoExplorer/Services/FileOperationService.cs
+++ b/PhotoGeoExplorer/Services/FileOperationService.cs
@@ -237,6 +237,14 @@ internal sealed class FileOperationService : IFileOperationService
                 continue;
             }
 
+            // 同一パスへのコピーは上書きするとソースを失うため無条件スキップ
+            if (IsSamePath(sourcePath, targetPath))
+            {
+                skipCount++;
+                progress?.Report(i + 1);
+                continue;
+            }
+
             if (ItemExistsAtPath(targetPath))
             {
                 ConflictResolution resolution;

